File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        count = 0
        ans = -1
        def dfs(root, k):
            if not root: return
            nonlocal count
            nonlocal ans
            dfs(root.left, k)
            print("count", count, "root.val", root.val)
            count += 1
            if count == k:
                ans = root.val
            dfs(root.right, k)
        
        dfs(root, k)
        return ans

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        if(count(nums.begin(), nums.end(), 0) >= 2){
            return vector<int>(nums.size(), 0);
        }
        int total = 1;
        for(int n:nums){
            if(n != 0)
                total*=n;
        }

        vector<int> ans;
        if(count(nums.begin(), nums.end(), 0) == 1){
            for(int i=0; i<nums.size(); i++){
                if(nums[i] == 0)
                    ans.push_back(total);
                else
                    ans.push_back(0);
            }
        }else{
            for(int i=0; i<nums.size(); i++){
                ans.push_back(total/nums[i]);
            }
        }
        return ans;
    }
};
